### PR TITLE
Publisher#flatMapConcatSingle null value concurrency fix

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Publisher#flatMapConcatSingle maintains a queue of elements
on top of Publisher#flatMapSingle to maintain ordering. Elements
within this queue have non-volatile state due to outer concurrency
protections, however there are two independent variables indicating
"terminal type" and "terminal value". This may result in reading
the "terminal type" the "terminal value" may not be visible yet
and return `null`.

Modifications:
- Collapse the 2 variables into a single variable, so when reading
  if it is non-null we have the "type" and "value" visible.

Result:
Fixes https://github.com/apple/servicetalk/issues/2321

### CODE REVIEW

**Comment by bondolo** [comment]
If the incomplete update can be observed then it must not be happening under the outer concurrency protection. Can multiple threads be racing to update the publisher or only one other thread observing the update? I am wondering if separate concurrency is needed on the publisher. Reducing to one variable will help, but I am uncertain that there still can't be update races--mostly null check followed by update rather than some flavor of compareAndSet`(null, value)`.

**Comment by Scottmitch** [comment]
The producer/consumer may happen on different threads. The consumer is single threaded (via lock) but the producer could be happening on another thread. If the producer is on another thread then the consumer may see partial state (if it is spread out over multiple variables). We can run through this offline as well to confirm assumptions.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java [hunk 1]
```diff
@@ -134,31 +136,27 @@ private void tryPollQueue() {
         SingleSource.Subscriber<? super R> subscriber;
         @Nullable
         private Object result;
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:138]: Consider adding a comment to clarify why it doesn't have to be `volatile`

**Review by idelpivnitskiy — APPROVED** [review]
Thanks for the quick fix!
While you are in RS-mode, please have a look at https://github.com/apple/servicetalk/pull/2320#discussion_r943757428 too.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java (hunk 1 above)
Comment (tkountis) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:138]: Yes I also think that this needs to be `volatile`. If not, what guarantees visibility from the producer to the consumer?

**Comment by Scottmitch** [comment]
discussed offline with @tkountis @bondolo @idelpivnitskiy . everyone is satisfied with the current approach and I added comments to clarify.

**Pull request merged**